### DISCUSSION TITLE
Using addIterativeLSI with a PeakMatrix

### DISCUSSION THREAD

**Discussion by skwalker** [GreenleafLab/ArchR#795] [Questions / Documentation]
Hi!

Thanks for building such a great platform for analyzing scATAC-seq data ! I've been playing around with using different values for `ceiling` when running `addPeakMatrix` and seeing how they change the lower dimensionality space by running `addIterativeLSI` with `useMatrix = "PeakMatrix"`. I've found that changing the value for `ceiling`, while it certainly changes the values in the Peak Matrix, does not seem to affect the lower dimensionality space; for example, my UMAP plots are exactly the same when I use `ceiling=1` and `ceiling=30`, and in addition, the SVD matrix generated from the reduced dimensions is also unchanged.  It seems a little bizarre to me that this wouldn't change anything since the Peak Matrix is quite different in these two situations. Do you have any idea of why updating the PeakMatrix like this wouldn't actually change the lower dimensionality space ? It is definitely using the PeakMatrix, right? 

Thanks,
Sarah

**Comment by rcorces** [comment]
I dont see a clear reason for how this would happen. Could you show some of the code that you have used?

**Reply by skwalker** [reply]
Thanks for the quick response!

So I'm just using my own dataset (I haven't tried with any of your publicly available datasets yet), but this is some of what I've done:
```
atacseq_proj_ceil1 = addPeakMatrix(atacseq_proj, ceiling = 1)
max(assay(getMatrixFromProject(atacseq_proj_ceil1, useMatrix="PeakMatrix"))) ## returns 1, as expected

atacseq_proj_ceil1 <- addIterativeLSI(
    ArchRProj = atacseq_proj_ceil1,
    useMatrix = "PeakMatrix", 
    name = "IterativeLSI", 
    iterations = 4, 
    clusterParams = list( 
        resolution = c(1.5), 
        sampleCells = 10000, 
        n.start = 10,
        maxClusters = 25
    ), 
    varFeatures = 52004, 
    dimsToUse = 1:30,
    force = TRUE,
    LSIMethod = 2,
)
ceil_1_lsi_mat <- atacseq_proj_ceil1@reducedDims$IterativeLSI$matSVD)

atacseq_proj_ceil30 = addPeakMatrix(atacseq_proj, ceiling = 30)
max(assay(getMatrixFromProject(atacseq_proj_ceil30, useMatrix="PeakMatrix"))) ## returns 30, as expected

atacseq_proj_ceil30 <- addIterativeLSI(
    ArchRProj = atacseq_proj_ceil30,
    useMatrix = "PeakMatrix", 
    name = "IterativeLSI", 
    iterations = 4, 
    clusterParams = list( 
        resolution = c(1.5), 
        sampleCells = 10000, 
        n.start = 10,
        maxClusters = 25
    ), 
    varFeatures = 52004, 
    dimsToUse = 1:30,
    force = TRUE,
    LSIMethod = 2,
)
ceil_30_lsi_mat <- atacseq_proj_ceil30@reducedDims$IterativeLSI$matSVD)
all.equal(ceil_1_lsi_mat, ceil_30_lsi_mat) ## returns TRUE, not expected
```

and if I do the basic steps of `addClusters`, `addUMAP` with the IterativeLSI as the `reducedDims`, and `plotEmbedding` with the UMAP as the embedding, I get identical UMAP plots.

**Reply by rcorces** [reply]
I'm not able to recapitulate your behavior (on ArchR v1.0.1). This is what I did, starting from `projHeme5` which already has a `PeakMatrix` and `IterativeLSI` run as shown in the tutorial:

```
proj <- saveArchRProject(ArchRProj = projHeme5, outputDirectory = "Save-ProjHeme5_b", load = TRUE)
max(assay(getMatrixFromProject(proj, useMatrix="PeakMatrix"))) ## returns 4, as expected

proj <- addPeakMatrix(proj, ceiling = 1)
max(assay(getMatrixFromProject(proj, useMatrix="PeakMatrix"))) ## returns 1, as expected

origLSI <- proj@reducedDims$IterativeLSI$matSVD

proj <- addIterativeLSI(
  ArchRProj = proj,
  useMatrix = "PeakMatrix", 
  name = "IterativeLSI", 
  iterations = 4, 
  clusterParams = list( 
    resolution = c(1.5), 
    sampleCells = 10000, 
    n.start = 10,
    maxClusters = 25
  ), 
  varFeatures = 52004, 
  dimsToUse = 1:30,
  force = TRUE,
  LSIMethod = 2,
)
newLSI <- proj@reducedDims$IterativeLSI$matSVD

all.equal(origLSI, newLSI)
# returns "Mean relative difference: 0.9247991"
```

So that begs the question, why are you getting different behavior? Could be how you are treating the ArchRProject and arrow files. Maybe create a copy of the project using `saveArchRProject()` and see if that helps? I dont see why it would cause this behavior but you are creating multiple ArchRProjects that point to the same arrow files which is always dicey.

**Reply by skwalker** [reply]
Hmm, this is odd. I am using ArchR version 1.0.1 and calling `saveArchRProject()` as you suggested still gives me different behavior than what you're seeing. Maybe I start from creating two separate `ArchRProject` from the `ArrowFiles` to see if that would prevent this weird state error ? 

I also tried setting different values for `name` when running `addIterativeLSI`, but that doesn't seem to be helping either.

I am also finding that `all.equal(getReducedDims(proj, reducedDims = "IterativeLSI_Ceiling30", returnMatrix = FALSE)$matSVD, proj@reducedDims$IterativeLSI_Ceiling30$matSVD)` returns `'Mean relative difference: 12.33672'
` which seems a little strange? Shouldn't those give you the same object ?

**Reply by rcorces** [reply]
Those operations are not equivalent because of scaling. From the `getReducedDims` documentation

scaleDims

> A boolean describing whether to z-score the reduced dimensions for each cell. This is useful for minimizing the contribution of strong biases (dominating early PCs) and lowly abundant populations. However, this may lead to stronger sample-specific biases since it is over-weighting latent PCs. If&nbsp;NULL&nbsp;this will scale the dimensions depending on if this were set true when the&nbsp;reducedDims&nbsp;were created by the dimensionality reduction method. This idea was introduced by Timothy Stuart.

since `addIterativeLSI()` defaults `scaleDims` to `TRUE`, `getReducedDims()` also scales whereas pulling

**Reply by skwalker** [reply]
Okay, that makes sense that the `getReducedDims()` would return a different matrix than just pulling the `reducedDims`  from the project itself.

Is it possible that the way the peak set is generated prior to building the peak matrix could affect the way the LSI reduction is done? I was also able to reproduce the correct behavior on the tutorial data, which is using `addReproduciblePeakSet` to generate the peak set. If I also use `addReproduciblePeakSet` on my dataset with Macs2,  I also get different results from the LSI dimensionality reduction when I run `addPeakMatrix` using different values for the `ceiling`. However, when I try to add my own peak set with `addPeakSet()`, the LSI dimensionality reduction remains the same (even though the PeakMatrix generated are quite different when generated the different values for `ceiling`, which is what I expect `addIterativeLSI` to be using).  So is there some way the functions `addPeakSet` and `addReproduciblePeakSet` do something different that would affect the results of `addIterativeLSI` when running on the PeakMatrix ?

**Reply by rcorces** [reply]
I'm still not able to reproduce the behavior you are talking about even when using `addPeakSet()`. My code is below. I think if you want to pursue this further, I'll need a full reproducible example based on the tutorial data.

```
projHeme5 <- loadArchRProject(path = "./Save-ProjHeme5/")
proj <- saveArchRProject(ArchRProj = projHeme5, outputDirectory = "Save-ProjHeme5_b", load = TRUE)

#extract the peak set, remove metadata and re-add the peak set using addPeakSet
repro_gr <- getPeakSet(projHeme5)
manual_gr <- repro_gr
mcols(manual_gr) <- NULL
proj <- addPeakSet(ArchRProj = proj, peakSet = manual_gr, force = TRUE)
getPeakSet(proj) #returns basic peak set without meta data as expected

proj <- addPeakMatrix(proj)
max(assay(getMatrixFromProject(proj, useMatrix="PeakMatrix"))) ## returns 4, as expected

proj <- addIterativeLSI(
  ArchRProj = proj,
  useMatrix = "PeakMatrix", 
  name = "IterativeLSI_1", 
  iterations = 4, 
  clusterParams = list( 
    resolution = c(1.5), 
    sampleCells = 10000, 
    n.start = 10,
    maxClusters = 25
  ), 
  varFeatures = 52004, 
  dimsToUse = 1:30,
  force = TRUE,
  LSIMethod = 2,
)
origLSI <- proj@reducedDims$IterativeLSI_1$matSVD

proj <- addPeakMatrix(proj, ceiling = 1)
max(assay(getMatrixFromProject(proj, useMatrix="PeakMatrix"))) ## returns 1, as expected

proj <- addIterativeLSI(
  ArchRProj = proj,
  useMatrix = "PeakMatrix", 
  name = "IterativeLSI_2", 
  iterations = 4, 
  clusterParams = list( 
    resolution = c(1.5), 
    sampleCells = 10000, 
    n.start = 10,
    maxClusters = 25
  ), 
  varFeatures = 52004, 
  dimsToUse = 1:30,
  force = TRUE,
  LSIMethod = 2,
)
newLSI <- proj@reducedDims$IterativeLSI_2$matSVD

all.equal(origLSI, newLSI)
# returns "Mean relative difference: 0.8213672"
```

**Reply by skwalker** [reply]
I've just copied your code and run it on my data, just changing the path for loading the original project, and it doesn't work for me (the line `all.equal(origLSI, newLSI)` returns true) with my data.

The project's PeakSet  for `getPeakSet(proj)`  still contains some metadata (idx, GC, and N) even though `manual_gr` of course doesn't have any metadata. Is that expected? I'm not sure why this would make a difference, but I am using the mouse mm10 genome (just `addArchRGenome("mm10")`).

**Reply by rcorces** [reply]
> The project's PeakSet for getPeakSet(proj) still contains some metadata (idx, GC, and N) even though manual_gr of course doesn't have any metadata. Is that expected?

Yes - when you add a peak set to your project, ArchR adds these metadata columns which are required for other functionality

Can you confirm whether the above code works for you on the tutorial data as written?

**Reply by skwalker** [reply]
Yes, it works on the tutorial data for me

**Reply by rcorces** [reply]
Ok that really has me stumped. I will try to look into this some more but this seems very odd.

**Reply by skwalker** [reply]
I think perhaps the issue might have something to do with the size of the peakset (my original one had 52k peaks)?

I was able to reiterate my original issue on the tutorial dataset by using a subset of the peak set extracted from the tutorial dataset (with `sample(peakset, 25000)`). Then, even though the peak matrices are quite different when using a ceiling of 1 and 16, the lower dimensionality spaces are exactly the same.

**Reply by rcorces** [reply]
Can you provide the exact code to reproduce this from the tutorial data?

**Reply by skwalker** [reply]
Ah I'm sorry for the delay  -- I thought I had already sent this message. Here is the code I used to show this:

```
library(ArchR)
set.seed(1)
addArchRThreads(threads = 1) 
inputFiles =   getTutorialData("Hematopoiesis")
addArchRGenome("hg19")

arrowfiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  filterTSS = 4, #Dont set this too high because you can always increase later
  filterFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = FALSE
)

proj <- ArchRProject(ArrowFiles = arrowfiles,
                     outputDirectory="archr_test_dir",
                     copyArrows=TRUE,)

# for adding clusters based on tiles
proj <- addIterativeLSI(
    ArchRProj = proj,
    useMatrix = "TileMatrix", 
    name = "IterativeLSI", 
    iterations = 2, 
    clusterParams = list( #See Seurat::FindClusters
        resolution = c(0.2), 
        sampleCells = 10000, 
        n.start = 10
    ), 
    varFeatures = 25000, 
    dimsToUse = 1:30
)

proj <- addClusters(
    input = proj,
    reducedDims = "IterativeLSI",
    method = "Seurat",
    name = "Clusters",
    resolution = 0.8
)

proj <- addGroupCoverages(ArchRProj = proj, groupBy = "Clusters")

# create peak set
pathToMacs2 <- findMacs2()
proj <- addReproduciblePeakSet(
    ArchRProj = proj, 
    groupBy = "Clusters",
    pathToMacs2 = pathToMacs2
)

proj = addPeakMatrix(proj, ceiling = 4)
max(assay(getMatrixFromProject(proj, useMatrix="PeakMatrix"))) # returns 4 

new_proj <- saveArchRProject(ArchRProj = proj, outputDirectory = "archr_test_dir_b", load = TRUE)
peakset <- getPeakSet(proj)

# create subsetted peak set
subsampled_peakset <- sample(peakset, 25000)
mcols(subsampled_peakset) <- NULL
subsampled_peakset #returns basic peak set with no meta data as expected

new_proj <- addPeakSet(ArchRProj = new_proj, peakSet = subsampled_peakset, force = TRUE)
getPeakSet(new_proj) #returns basic peak set with some meta data

# dimensionality reduction with ceiling of 1 for peak matrix
new_proj <- addPeakMatrix(new_proj, ceiling = 1)
max(assay(getMatrixFromProject(new_proj, useMatrix="PeakMatrix"))) ## returns 1, as expected

new_proj <- addIterativeLSI(
  ArchRProj = new_proj,
  useMatrix = "PeakMatrix", 
  name = "IterativeLSICeiling1", 
  iterations = 4, 
  clusterParams = list( 
    resolution = c(1.5), 
    sampleCells = 10000, 
    n.start = 10,
    maxClusters = 25
  ), 
  varFeatures = 52004, 
  dimsToUse = 1:30,
  force = TRUE,
  LSIMethod = 2,
)
ceiling1LSI <- new_proj@reducedDims$IterativeLSICeiling1$matSVD


# dimensionality reduction with ceiling of 13 for peak matrix
new_proj <- addPeakMatrix(new_proj, ceiling=13)
max(assay(getMatrixFromProject(new_proj, useMatrix="PeakMatrix"))) ## returns 13, as expected

new_proj <- addIterativeLSI(
  ArchRProj = new_proj,
  useMatrix = "PeakMatrix", 
  name = "IterativeLSICeiling13", 
  iterations = 4, 
  clusterParams = list( 
    resolution = c(1.5), 
    sampleCells = 10000, 
    n.start = 10,
    maxClusters = 25
  ), 
  varFeatures = 52004, 
  dimsToUse = 1:30,
  force = TRUE,
  LSIMethod = 2,
)
ceiling13LSI <- new_proj@reducedDims$IterativeLSICeiling13$matSVD

all.equal(ceiling1LSI, ceiling13LSI) # returns TRUE, not as expected
```

**Reply by rcorces** [reply]
I can confirm that I am able to recapitulate this behavior using the code you provided. I am also able to recapitulate this behavior when creating two new projects from a blank-slate tutorial project that lacks any PeakMatrix, meaning that I dont think this behavior has anything to do with project subsetting, weird issues with saveArchRProject, or issues with overwriting a peak matrix etc.

That being said, I havent yet figured out what the problem is because the actual PeakMatrix objects are indeed different when using different values for ceiling. I'll keep looking into this.

**Reply by rcorces** [reply]
I'm moving this to an Issue for easier tracking. Please follow https://github.com/GreenleafLab/ArchR/issues/941 and post there for followup
